test_graph.py

```diff
@@ -142,14 +142,53 @@ def test_adjacent():
     graph.add_edge(20, 6)
 
     assert graph._storage == {
-        4: {(4, 6), (4, 9)},
-        6: {(6, 4), (6, 20)},
-        9: {(9, 4)},
-        20: {(20, 6)},
+        4: {6, 9},
+        6: {4, 20},
+        9: {4},
+        20: {6},
     }
     assert graph.adjacent(6, 4) == True
     assert graph.adjacent(20, 4) == False
     assert graph.adjacent(4, 6) == True
     with pytest.raises(ValueError) as exc_info:
         graph.adjacent(4, 15)
     assert exc_info.value.args[0] == "One of the values are not in the graph"
+
+
+def test_depth_first_traversal():
+    graph = Graph()
+    graph.add_edge(4, 6)
+    graph.add_edge(4, 9)
+    graph.add_edge(4, 12)
+    graph.add_edge(6, 20)
+    graph.add_edge(6, 13)
+    graph.add_edge(13, 7)
+    graph.add_edge(9, 28)
+
+    assert graph.depth_first_traversal(4) == [4, 6, 20, 13, 7, 12, 9, 28]
```

Comment:
This is a challenging thing to test.  Sets are not ordered. Like dictionaries.  You path is derived from the order you loop over the Set.  I'm okay with you leaving it like this, but maybe think a moment about how you could write these to not depend on the order of the sets.  For example if you were to create the Graph in 2 different ways, the tests should still pass, like:
```suggestion
def test_depth_first_traversal():
    graph = Graph()
    graph.add_edge(4, 6)
    graph.add_edge(4, 9)
    graph.add_edge(4, 12)
    graph.add_edge(6, 20)
    graph.add_edge(6, 13)
    graph.add_edge(13, 7)
    graph.add_edge(9, 28)

    # assert something that's always true
    
    graph = Graph()
    graph.add_edge(4, 12)
    graph.add_edge(4, 9)
    graph.add_edge(4, 6)
    graph.add_edge(6, 20)
    graph.add_edge(6, 13)
    graph.add_edge(13, 7)
    graph.add_edge(9, 28)
    
    # assert the same always true thing
```
I just changed the order that the edges were added, but the graph is the same so ideally it would test out the same. Again, you don't have to change these, but think on it some.

Reply:
could i just change my edges to be lists instead? that way i can guarantee the order?

Reply:
You could, but you'ed be making a design choice based on how easy it is to test.  What side effects would that have to how the code operates.  Can you think of any downside to having them be lists?

Reply:
/i would increase the run time any time you access the edges?

Reply:
Right, specifically [here](https://github.com/aepaysinger/data-structures/pull/10/files#diff-bfd5ce89c947f4ba81d03b804ba27d0fdc8686974b21bbbf9e37401080a67817R70) [here](https://github.com/aepaysinger/data-structures/pull/10/files#diff-bfd5ce89c947f4ba81d03b804ba27d0fdc8686974b21bbbf9e37401080a67817R49) and [here](https://github.com/aepaysinger/data-structures/pull/10/files#diff-bfd5ce89c947f4ba81d03b804ba27d0fdc8686974b21bbbf9e37401080a67817R42)

Reply:
could i use and ordered set instead? doesnt that keep items in the same position?

Reply:
Python doesn't include an ordered set. It possible to make one and also there are libraries that do include it, but both are beyond the scope of this exercise.

Reply:
okay so i changed it to have another loop that sorts it. that way no matter what its always putting in the smallest number first